File: phoenix/lib/cards_project/cards/deck.ex
defmodule CardsProject.Cards.Deck do
  use Ecto.Schema
  import Ecto.Changeset

  schema "decks" do
    field :name, :string
    field :description, :string
    field :format, :string
    field :is_public, :boolean, default: false
    field :is_tournament_legal, :boolean, default: false
    field :archetype, :string
    field :wins, :integer, default: 0
    field :losses, :integer, default: 0
    field :draws, :integer, default: 0
    field :created_at, :naive_datetime
    belongs_to :player, CardsProject.Players.Player
    many_to_many :cards, CardsProject.Cards.Card, join_through: "deck_cards"
    many_to_many :sideboard_cards, CardsProject.Cards.Card, join_through: "deck_sideboard_cards"
    many_to_many :tags, CardsProject.Cards.DeckTag, join_through: "deck_tag_assignments"
    has_many :deck_cards, CardsProject.Cards.DeckCard, foreign_key: :deck_id
    has_many :tag_assignments, CardsProject.Cards.DeckTagAssignment, foreign_key: :deck_id
    has_many :tournament_registrations, CardsProject.Tournaments.TournamentRegistration, foreign_key: :deck_id
    has_many :articles, CardsProject.Content.Article, foreign_key: :featured_deck_id

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:name, :is_public, :is_tournament_legal, :wins, :losses, :draws, :created_at, :description, :format, :archetype, :player_id])
    |> validate_required([:name, :is_public, :is_tournament_legal, :wins, :losses, :draws, :created_at])
    |> validate_inclusion(:format, ["Standard", "Extended", "Legacy", "Vintage", "Commander", "Draft"])
    |> validate_inclusion(:archetype, ["Aggro", "Control", "Midrange", "Combo", "Prison", "Tempo"])
    |> validate_number(:wins, greater_than_or_equal_to: 0, message: "Deck wins count must not be negative")
    |> validate_number(:losses, greater_than_or_equal_to: 0, message: "Deck losses count must not be negative")
    |> validate_number(:draws, greater_than_or_equal_to: 0, message: "Deck draws count must not be negative")
    |> then(fn cs ->
      if get_field(cs, :is_tournament_legal) == "true" and (get_field(cs, :is_public) != "true") do
        Ecto.Changeset.add_error(cs, :is_public, "Tournament-legal deck must be made public")
      else
        cs
      end
    end)
  end

  @doc false
  def update_changeset(record, attrs) do
    record
    |> cast(attrs, [:name, :is_public, :is_tournament_legal, :description, :format, :archetype, :player_id])
    |> validate_required([:name, :is_public, :is_tournament_legal])
  end

  # ── Business operations ────────────────────────────────────────────

  def validate_size(_record) do
    # TODO: implement Deck.validate_size
    {:error, :not_implemented}
  end

  def add_card(_record, _card_id, _quantity) do
    # TODO: implement Deck.add_card
    :ok
  end

  def remove_card(_record, _card_id) do
    # TODO: implement Deck.remove_card
    :ok
  end

  def win_rate(_record) do
    # TODO: implement Deck.win_rate
    {:error, :not_implemented}
  end

  def clone(_record) do
    # TODO: implement Deck.clone
    {:error, :not_implemented}
  end

  def publish(_record) do
    # TODO: implement Deck.publish
    :ok
  end

  def unpublish(_record) do
    # TODO: implement Deck.unpublish
    :ok
  end

  def certify_tournament_legal(_record) do
    # TODO: implement Deck.certify_tournament_legal
    {:error, :not_implemented}
  end
end

File: laravel/app/Models/Cards/Deck.php
<?php

namespace App\Models\Cards;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use App\Models\Players\Player;

class Deck extends Model
{
    protected $table = 'decks';

    protected $fillable = ['name', 'description', 'format', 'is_public', 'is_tournament_legal', 'archetype', 'wins', 'losses', 'draws', 'player_id'];

    protected $casts = [
        'is_public' => 'boolean',
        'is_tournament_legal' => 'boolean',
    ];

    const FORMAT_VALUES = ['Standard', 'Extended', 'Legacy', 'Vintage', 'Commander', 'Draft'];
    const ARCHETYPE_VALUES = ['Aggro', 'Control', 'Midrange', 'Combo', 'Prison', 'Tempo'];

    public function player(): BelongsTo
    {
        return $this->belongsTo(Player::class, 'player_id');
    }

    public function cardses(): BelongsToMany
    {
        return $this->belongsToMany(Card::class, 'deck_cards_pivot', 'deck_id', 'card_id');
    }

    public function sideboardCardses(): BelongsToMany
    {
        return $this->belongsToMany(Card::class, 'deck_sideboard_cards_pivot', 'deck_id', 'card_id');
    }

    public function tagses(): BelongsToMany
    {
        return $this->belongsToMany(DeckTag::class, 'deck_tags_pivot', 'deck_id', 'deck_tag_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->wins === null || $this->wins >= 0))) {
            $errors['wins_not_negative'] = 'Deck wins count must not be negative';
        }
        if (!(($this->losses === null || $this->losses >= 0))) {
            $errors['losses_not_negative'] = 'Deck losses count must not be negative';
        }
        if (!(($this->draws === null || $this->draws >= 0))) {
            $errors['draws_not_negative'] = 'Deck draws count must not be negative';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->is_tournament_legal === true && !($this->is_public === true)) {
            throw new \RuntimeException('Tournament-legal deck must be made public');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function validateSize(): bool
    {
        throw new \RuntimeException('validate_size not implemented');
    }

    public function addCard($card_id, $quantity): void
    {
        throw new \RuntimeException('add_card not implemented');
    }

    public function removeCard($card_id): void
    {
        throw new \RuntimeException('remove_card not implemented');
    }

    public function winRate(): string
    {
        throw new \RuntimeException('win_rate not implemented');
    }

    public function clone(): mixed
    {
        throw new \RuntimeException('clone not implemented');
    }

    public function publish(): void
    {
        throw new \RuntimeException('publish not implemented');
    }

    public function unpublish(): void
    {
        throw new \RuntimeException('unpublish not implemented');
    }

    public function certifyTournamentLegal(): bool
    {
        throw new \RuntimeException('certify_tournament_legal not implemented');
    }

}
